lib/codecs/src/decoding/format/gelf.rs

```diff
@@ -0,0 +1,216 @@
+            // Libraries SHOULD not allow to send id as additional field ( _id). Graylog server nodes omit this field automatically.
+            .unknown_fields(Kind::bytes())
+        //
+    }
+}
+
+/// Deserializer that builds an `Event` from a byte frame containing a GELF log
+/// message.
+#[derive(Debug, Clone)]
+pub struct GelfDeserializer {
+    regex: Regex,
+}
+
+impl Default for GelfDeserializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GelfDeserializer {
+    /// Create a new GelfDeserializer
+    pub fn new() -> GelfDeserializer {
+        GelfDeserializer {
+            regex: Regex::new(r"^[\w\.\-]*$").unwrap(),
+        }
+    }
+
+    // Builds a LogEvent from the parsed GelfMessage, adhering to GELF spec
+    fn insert_fields_from_gelf(&self, parsed: &GelfMessage) -> vector_core::Result<Event> {
+        let mut event = Event::from(parsed.short_message.to_string());
+        let log = event.as_mut_log();
+
+        // GELF spec defines the version as 1.1 which has not changed since 2013
+        if parsed.version != GELF_VERSION {
+            return Err(format!(
+                "{} does not match GELF spec version ({})",
+                VERSION, GELF_VERSION
+            )
+            .into());
+        }
+
+        log.insert(VERSION, parsed.version.to_string());
+        log.insert(HOST, parsed.host.to_string());
+
+        if let Some(full_message) = &parsed.full_message {
+            log.insert(FULL_MESSAGE, full_message.to_string());
+        }
+
+        if let Some(timestamp) = parsed.timestamp {
+            let naive = NaiveDateTime::from_timestamp(
+                f64::trunc(timestamp) as i64,
+                f64::fract(timestamp) as u32,
+            );
+            log.insert(
+                log_schema().timestamp_key(),
+                DateTime::<Utc>::from_utc(naive, Utc),
+            );
+        } else {
+            log.insert(log_schema().timestamp_key(), Utc::now());
+        }
+
+        if let Some(level) = parsed.level {
+            log.insert(LEVEL, level);
+        }
+        if let Some(facility) = &parsed.facility {
+            log.insert(FACILITY, facility.to_string());
+        }
+        if let Some(line) = &parsed.line {
+            log.insert(LINE, *line);
+        }
+        if let Some(file) = &parsed.file {
+            log.insert(FILE, file.to_string());
+        }
+
+        if let Some(add) = &parsed.additional_fields {
+            for (key, val) in add.iter() {
+                // per GELF spec, filter out _id
+                if key == "_id" {
+                    continue;
+                }
+                // per GELF spec, Additional field names must be characters dashes or dots
+
+                // drop fields names that don't match the GELF spec
+                if !self.regex.is_match(key) {
+                    continue;
+                }
```

Comment:
Do you know what Graylog itself does with invalid fields? Does it just drop those fields or does it reject the whole line? It might be good to match their behavior since that is likely to be what users expect.

It looks like there is an open source version you can run to test: https://www.graylog.org/products/open-source

Reply:
Great suggestion- I've found some interesting discrepancies from the spec and the graylog behavior. Essentially they are much more flexible than the spec suggests. I'll be updating the logic to be more flexible.

Reply:
In case anyone is curious, these are my findings from experimenting with the graylog product:
```
// if version is not 1.1 or is omitted, it doesn't care, or if it is not a string, doesn't care

// if host is not specified, it uses the IP address , if host is not a string  it drops the event

//  if short_message is missing , it drops the event

// if timestamp is the wrong type, it uses the current time

//  (level and line) if is a string but numeric, it parses it . if level is a non numeric string, it drops event
// (level and line) if = float , it rounds down


// (facility and file) can be any type

// if invalid character in field name, it drops that field

// if additional field not underscored , it doesn't care

// if additional field name has underscore, it trims the underscore
```